docs/ffi.md

```diff
@@ -493,13 +501,174 @@ print(jax.jit(rms_norm_cross_platform).lower(x).as_text(dialect="hlo").strip())
+```
+
+But, if you look at the compiled HLO (omitting a helper functions for clarity), you'll see that
+
+1. the data are first fully replicated onto each device via an `all-gather` operation,
+2. the FFI call is executed on the full dataset on each device, and
+3. the output is sliced to discard the unused portions.
+
+```{code-cell} ipython3
+hlo = jax.jit(rms_norm, out_shardings=batch_shd).lower(x_batch_shd).compile().as_text().strip()
+print(hlo.split("\n\n")[-1])
+```
+
+This clearly (to us!) isn't the optimal partitioning of this function, but it's the best that JAX/XLA can do with the information given.
+
+To generate better partitioning logic, we can use {func}`~jax.experimental.shard_map.shard_map` or {func}`~jax.experimental.custom_partitioning.custom_partitioning`, and we discuss both options here.
+That being said, it's not straightforward to generate _optimal_ partitioning for all inputs, because sometimes this would require algorithmic changes.
+Specifically, let's add support for "batch partitioning", which handles the case where the data are sharded on batch dimensions, but sharding on the last dimension will always require in re-sharding.
+
+### Using `shard_map`
+
+If you are using manual sharding control via {func}`~jax.experimental.shard_map.shard_map`, any FFI calls in your program should already partition appropriately:
+
+```{code-cell} ipython3
+from functools import partial
+from jax.experimental.shard_map import shard_map
+
+@partial(shard_map, mesh=mesh, in_specs=P("x", None), out_specs=P("x", None))
+def rms_norm_shmap(x):
+  return rms_norm(x)
+
+np.testing.assert_allclose(rms_norm_shmap(x_batch_shd), rms_norm_ref(x), rtol=1e-5)
+print(jax.jit(rms_norm_shmap, out_shardings=batch_shd).lower(x_batch_shd).compile().as_text().strip())
+```
+
+As you can see in this program, if the input and output shardings match the `shard_map` specs, no communication is required and the FFI call is executed on the appropriately sharded subset of the data.
+
+You can also use inputs and outputs with shardings that don't match the `shard_map` specs, but (unrelated to the FFI) this will require re-sharding, as seen by the `all-to-all` operations in the compiled HLO:
+
+```{code-cell} ipython3
+hlo_data_shmap = jax.jit(rms_norm_shmap, out_shardings=data_shd).lower(x_data_shd).compile().as_text()
+assert "all-to-all" in hlo_data_shmap
+```
+
+### Using `custom partitioning`
+
+If you can't use {func}`~jax.experimental.shard_map.shard_map`, an alternative approach is to use {func}`~jax.experimental.custom_partitioning.custom_partitioning`, which supports automatic parallelization via {func}`jax.jit`.
+{func}`~jax.experimental.custom_partitioning.custom_partitioning` works by adding Python callbacks into the XLA compiler's partitioning pass, which allows very flexible logic, but also comes with some rough edges.
+We won't go into too much detail on the caveats here, but the main issues that you should be aware of are:
```

Comment:
Should these be documented separately (apologies if they already are and I missed that)?

Reply:
Good question! No, I don't think so. I also don't think this is necessarily the right place to document all of this, but I thought it was important to comment on these specific issues because I hit them when writing up this tutorial. I'm inclined to leave this as is, and hopefully things will get a little cleaner with the shardy approach in the future, or something else!